narwhal/primary/src/state_handler.rs

```diff
@@ -1,35 +1,185 @@
     rx_committed_certificates: Receiver<(Round, Vec<Certificate>)>,
     /// Channel to signal committee changes.
     rx_shutdown: ConditionalBroadcastReceiver,
     /// A channel to update the committed rounds
     tx_committed_own_headers: Option<Sender<(Round, Vec<Round>)>>,
+    /// A channel to send system messages to the proposer.
+    tx_system_messages: Sender<SystemMessage>,
+
+    /// If set, generates Narwhal system messages for random beacn
+    /// DKG and randomness generation.
+    randomness_state: Option<RandomnessState>,
 
     network: anemo::Network,
 }
 
+// Internal state for randomness DKG and generation.
+struct RandomnessState {
+    party: dkg::Party<PkG, EncG>,
+    messages: Vec<dkg::Message<PkG, EncG>>,
+    processed_messages: Vec<dkg::ProcessedMessage<PkG, EncG>>,
+    shares: dkg::SharesMap<<PkG as groups::GroupElement>::ScalarType>,
+    confirmations: Vec<dkg::Confirmation<EncG>>,
+    dkg_output: Option<dkg::Output<PkG, EncG>>,
+}
+
+impl RandomnessState {
+    fn try_new(
+        protocol_config: ProtocolConfig,
+        committee: Committee,
+        private_key: RandomnessPrivateKey,
+    ) -> Option<Self> {
+        if !protocol_config.random_beacon() {
+            return None;
+        }
+
+        let info = committee.randomness_dkg_info();
+        let nodes = info
+            .iter()
+            .map(|(id, pk, stake)| nodes::Node::<EncG> {
+                id: id.0,
+                pk: pk.clone(),
+                weight: *stake as u16,
+            })
+            .collect();
+        let nodes = match nodes::Nodes::new(nodes) {
+            Ok(nodes) => nodes,
+            Err(err) => {
+                error!("Error while initializing random beacon state: {err:?}");
+                return None;
+            }
+        };
+        let (nodes, t) = nodes.reduce(
+            VALIDITY_THRESHOLD
+                .try_into()
+                .expect("VALIDITY_THRESHOLD should fit in u16"),
+            protocol_config.random_beacon_reduction_allowed_delta(),
+        );
+        let party = match dkg::Party::<PkG, EncG>::new(
+            private_key,
+            nodes,
+            t.into(),
+            fastcrypto_tbls::random_oracle::RandomOracle::new(
+                format!("dkg {}", committee.epoch()).as_str(),
+            ),
+            &mut rand::thread_rng(),
+        ) {
+            Ok(party) => party,
+            Err(err) => {
+                error!("Error while initializing random beacon state: {err:?}");
+                return None;
+            }
+        };
+        Some(Self {
+            party,
+            messages: Vec::new(),
+            processed_messages: Vec::new(),
+            shares: dkg::SharesMap::with_capacity(0),
+            confirmations: Vec::new(),
+            dkg_output: None,
+        })
+    }
+
+    async fn start_dkg(&self, tx_system_messages: &Sender<SystemMessage>) {
+        let msg = self.party.create_message(&mut rand::thread_rng());
```

Comment:
Maybe the rng could be injected during the construction of the struct and used across? That could also help during testing to ensure some determinism if needed

Reply:
I think it's simplest to keep an extra field out of the RandomnessState (which as you note is already a bit intricate) until there is a reason we need it?